### PR TITLE
feat: generate strategies intermediate results

### PR DESCRIPTION
**Pull request by Carsopre**
## Issue addressed
Solves #194

## Code of conduct
- [x] I HAVE NOT added sensitive or compromised (test) data to the repository.
- [x] I HAVE NOT added vulnerabilities to the repository.
- [x] I HAVE discussed my solution with (other) members of the KOSWAT team.

## What has been done?
- Updated the current dataclass definition of `StrategyLocationReinforcement` so that we can keep track of the different steps added throughout the different strategies.
- Added new `StrategyStepEnum` and `StrategyStepAssignment` so that it's easier to track said steps.
- Small correction to `OrderStrategy.get_strategy_reinforcements` so that the generated reinforcements have their `StrategyLocationReinforcement.available_measures` in the same order as the one set from the `OrderStrategy`.
- Modified `SummaryLocationsShpExporter to create a new `ClusterCollectionShpFom` for the "step" layer.
- Extended and extracted `ClusterCollectionShpFom` by creating a dataclass to represent its geodataframes output ( `ClusterGeodataframeOutputFom`).
- **Important!** Modified `KoswatSummaryBuilder` with `strategy_type` property that allows us to indicate which strategy needs to be run (by default `StrategyOrder`.
- Overall update of documentation.

### Checklist
- [x] Tests are either added or updated.
- [x] Branch is up to date with `master`.
- [x] Updated documentation if needed.

## Additional Notes (optional)
- Tests reference data has been updated as the analysis generates now new columns for the locations csv output.
- Important! `TestAcceptance::test_koswat_run_as_sandbox_with_obstacles_and_infrastructures` runs now with `InfraPriorityStrategy`

### CODE REVIEW

**Review by ArdtK — CHANGES_REQUESTED** [review]

File: koswat/strategies/strategy_location_reinforcement.py [hunk 1]
```diff
@@ -8,30 +8,120 @@
...
+        # Update the assigned value for this step.
+        self._selected_measure_steps.append(
+            StrategyStepAssignment(
+                step_number=len(self._selected_measure_steps),
+                step_type=step,
+                step_value=reinforcement_type,
+            )
+        )
+
+    def get_selected_measure_steps(
```
[the remark this answers is not in the document]
Comment (ArdtK) [reply koswat/strategies/strategy_location_reinforcement.py:91]: Already discussed

File: docs/reference/koswat_cost_report.md [hunk 2]
```diff
@@ -84,11 +84,16 @@ After running a Koswat analysis, several files and directories will be generated
 - Dike section: The selected dike section being analyzed.
 - Generated files: A combination of images and a 'csv' matrix result.
     - Images: Visual description of each of the possible reinforcements being applied.
-    - summary_costs.csv: A csv file containing all the costs information of the summary.
+    - `summary_costs.csv`: A csv file containing all the costs information of the summary.
         - Represents the Summary, Profile and Layer report.
-    - summary_locations.csv: A csv file containing per-location a breakdown of available reinforcements and selected reinforcement ( see [strategies](koswat_strategies.md)).
+    - `summary_locations.csv` (phased out): A csv file containing per-location a breakdown of available reinforcements and selected reinforcement ( see [strategies](koswat_strategies.md)).
```
[the remark this answers is not in the document]
Comment (Carsopre) [reply docs/reference/koswat_cost_report.md:89]: The original idea of the `.shp` is indeed to have the information there, in principle we keep this `.csv` to validate the early phases of our implementation.

File: docs/reference/koswat_cost_report.md [hunk 3]
```diff
@@ -84,11 +84,16 @@ After running a Koswat analysis, several files and directories will be generated
...
         - Represents the Location report.
-    - summary_infrastructure_costs.csv: A csv file contaning all the infrastructure costs at each location for each of the supported reinforcement profile types.
+    - `/summary_locations` directory: contains the following `.shp` files (and their related binaries):
+        -`summary_locations_measures`: The same data as present in `summary_locations.csv` is used to shape the geometry of a dike's traject. We keep as well the type of chosen reinforcement.
+        - `summary_locations_new`: an overlay of `summary_locations_measures` with the geometry buffered in relation to the new selected reinforcement.
```
[the remark this answers is not in the document]
Comment (Carsopre) [reply docs/reference/koswat_cost_report.md:93]: This is based on the code from peter and is reflected in the description here and in the code:
- `_new`, the new selected reinforcement (or final if you wish)
- `_old`, original selection reinforcement (or initial if you wish)

File: docs/reference/koswat_cost_report.md [hunk 4]
```diff
@@ -84,11 +84,16 @@ After running a Koswat analysis, several files and directories will be generated
...
+        - `summary_locations_old`: an overlay of `summary_locations_measures` with the geometry buffered in relation to the original selected reinforcement.
+        - `summary_locations_step`: an overlay of `summary_locations_measures` with the geometry buffered in relation to the [Ordered Strategy](koswat_strategies.md#order-based-default) step.
```
[the remark this answers is not in the document]
Comment (Carsopre) [reply docs/reference/koswat_cost_report.md:95]: No, the step is the output of the order based. The output from infrastructure is the final result ( `_new`) which could of course be exactly the same as the one from step

File: koswat/strategies/strategy_step/strategy_step_assignment.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+from koswat.dike_reinforcements.reinforcement_profile.reinforcement_profile_protocol import (
+    ReinforcementProfileProtocol,
+)
+from koswat.strategies.strategy_step.strategy_step_enum import StrategyStepEnum
+
+
+@dataclass
+class StrategyStepAssignment:
+    step_number: int
```
[the remark this answers is not in the document]
Comment (Carsopre) [reply koswat/strategies/strategy_step/strategy_step_assignment.py:11]: oh yes I forgot about this as I was going through different implementations.

File: koswat/strategies/order_strategy/order_strategy.py [hunk 6]
```diff
@@ -117,21 +118,42 @@ def get_strategy_reinforcements(
...
+            # Create the strategy representation.
+            _slr = StrategyLocationReinforcement(
+                location=_strategy_location.point_surrounding,
+                available_measures=_available_measures,
+                strategy_location_input=_strategy_location,
+            )
+
+            # Manually set the current selected measure if none available.
+            if not any(_slr.available_measures):
+                # TODO: Unclear why tests fail when setting this measure
```
[the remark this answers is not in the document]
Comment (Carsopre) [reply koswat/strategies/order_strategy/order_strategy.py:152]: This does not come from now, I reported it last week, It's just simply moved.

File: tests/strategies/test_strategy_location_reinforcement.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
+from typing import Iterator
+
+import pytest
+
+from koswat.dike_reinforcements.reinforcement_profile.outside_slope.cofferdam_reinforcement_profile import (
+    CofferdamReinforcementProfile,
+)
```
[the remark this answers is not in the document]
Comment (Carsopre) [reply tests/strategies/test_strategy_location_reinforcement.py:7]: From the test perspective I'd rather make an explicit import (and I try to change it whenever I see it). This is also the automatic import you get from `vscode`.
From the source code I agree (as long as it's referencing from "sibbling" projects.

File: tests/test_data/acceptance_reference_data/case_dijk1_1a_dh_with_default_layers/summary_locations.csv
[the remark this answers is not in the document]
Comment (Carsopre) [reply tests/test_data/acceptance_reference_data/case_dijk1_1a_dh_with_default_layers/summary_locations.csv:1]: What I talked with Peter was to name the last column as Optimized. We can reduce the number of columns later on, for now I'm not going to modify it as the code and functionality impact is minimal and we want to merge everything by today.

File: tests/strategies/test_strategy_location_reinforcement.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ArdtK) [reply tests/strategies/test_strategy_location_reinforcement.py:7]: I would argue that it's a kind of encapsulation of your inner organization of your sub-projects.

File: tests/test_acceptance.py
[the remark this answers is not in the document]
Comment (ArdtK) [reply tests/test_acceptance.py:1]: Discussed and will continue with @peterdgr

File: docs/reference/koswat_cost_report.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ArdtK) [reply docs/reference/koswat_cost_report.md:93]: I would prefer to be explicit, like "initial", "infra strategy"

File: docs/reference/koswat_cost_report.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArdtK) [reply docs/reference/koswat_cost_report.md:89]: As I assume we will maintain it, I'd rather remove the "phased out"

**Review by ArdtK — APPROVED** [review]

**Pull request merged**